test/box-luatest/gh_10142_box_ctl_wal_sync_test.lua

```diff
@@ -0,0 +1,82 @@
+local t = require('luatest')
+local server = require('luatest.server')
+
+local g = t.group()
+
+g.before_all(function(g)
+    g.server = server:new({alias = 'master'})
+    g.server:start()
+end)
+
+g.after_all(function(g)
+    g.server:drop()
+end)
+
+g.after_each(function(g)
+    g.server:exec(function()
+        if box.space.test ~= nil then
+            box.cfg{read_only=false}
+            box.space.test:drop()
+        end
+    end)
+end)
+
+g.test_box_ctl_val_sync = function(g)
+    g.server:exec(function()
+        local fiber = require('fiber')
+        t.tarantool.skip_if_not_debug()
+        -- Let's follow proposed 10142 reproducer as close as possible
+        -- Step 1: Start a Tarantool built in debug mode
+        box.cfg{}
+
+        -- Step 2: Create a space
+        local s = box.schema.space.create('test')
+        s:format({{ name = 'pk', type = 'unsigned' },})
+        s:create_index('primary', { parts = { 'pk' } })
+
+        -- Step 3.1: Insert wal delay so it's possible to commit
+        -- but not writing to the wal affecting lsn
+        box.error.injection.set('ERRINJ_WAL_DELAY', true)
+        -- Step 3.2: Save initial lsn value
+        local lsn_before = box.info.lsn
+
+        local lsn_save = lsn_before
+        local lsn_save_finished = false
+        local lsn_insert_started = false
+        local lsn_insert_finished = false
+        fiber.new(function()
+            -- Step 4: Initiate long write which we need to
+            -- wait for some reason
+            lsn_insert_started = true
+            box.space.test:insert{1}
+            lsn_insert_finished = true
+        end)
+        fiber.new(function()
+            while not lsn_insert_started do
+                fiber.yield()
+            end
+            -- Step 5
+            fiber.yield() -- why not yield here?
+            box.ctl.wal_sync()
+            -- Step 6: Save the result lsn
+            lsn_save = box.info.lsn
+
+            lsn_save_finished = true
+        end)
+        -- Step 7: re-enable wal again so write can happen
+        -- now
+        box.error.injection.set('ERRINJ_WAL_DELAY', false)
+        while not lsn_insert_finished do
+            fiber.yield()
+        end
+        -- Step 8: lsn is guaranteed to be after that transaction
+        local lsn_after = box.info.lsn
+        while not lsn_save_finished do
```

Comment:
It's always better to use:

```lua 
t.helpers.retrying({timeout = ...}, function()
    t.assert_equals(lsn_save_finished, true)
end)
```

Reply:
Isn't it a problem to add sleeps in tests? But this was actually convenient, thanks. Fixed

Reply:
> Isn't it a problem to add sleeps in tests?

We generally try to avoid doing so. It's always better to wait for some condition to happen than adding some incomprehensible `fiber.sleep`